Added --sort option support to bossquery.py, taking a SQL formatted s…

dcunning11235 opened the pull request:
…ort string

Added --sort option support to meta.py Database.select_all method

Altered meta.py create_meta_lite method to parse only first 100 lines using `astropy` methods, then read in remainder using numpy.loadtxt (per dkirkby's comment down-thread)

Updated bin/scripts.rst with additional information about command line args to bossquery

dcunning11235 committed:
Added --sort option support to bossquery.py, taking a SQL formatted sort string
Added --sort option support to meta.py Database.select_all method

Altered meta.py create_meta_lite method to parse and insert records in chunks.
This defaults to 50k, but is setable via the environment variable BOSS_CHUNK_SIZE

This does NOT affect create_meta_full

dkirkby commented on bin/bossquery:
```diff
@@ -32,11 +32,13 @@ def main():
     parser.add_argument('--no-clobber', action = 'store_true',
         help = 'Do not overwrite any existing save file with the same name.')
+    parser.add_argument('--sort', type = str, default = None, metavar = 'SORT',
+        help = 'Sort; follows SQL syntaxt, columns listed in order of sort priority, DESC optionally specified to reverse order.')
```
Typo: "syntaxt"

dkirkby commented on bossdata/meta.py:
```diff
@@ -98,36 +101,68 @@ def create_meta_lite(sp_all_path, db_path, verbose=True):
     """
+    
+    #parameters controlling how parse/inserts are chunked
+    chunk_size_str = os.getenv('BOSS_CHUNK_SIZE')
```
I would like to avoid adding another envvar if possible, especially for an implementation parameter (rather than a user config parameter).  Will a fixed value work, or can you code the logic for setting a sensible value?

dkirkby commented on bossdata/meta.py:
```diff
@@ -98,36 +101,68 @@ def create_meta_lite(sp_all_path, db_path, verbose=True):
+            cursor.executemany(sql, table)
+            
+            if verbose:
+                print(position+chunk_size-init_position, "rows @", datetime.datetime.now())
```
This is probably more detailed than most users want. Can you use progress_bar instead here?

dkirkby commented:
@dcunning11235 Please add some text to `bin/scripts.rst` describing the new `--sort` option (and any other updates you think are necessary).

dcunning11235 committed:
Updates per comments from @dkirkby:

- Fixed message typo in bin/bossquery

- Removed BOSS_CHUNK_SIZE and hardcoded chunk to be 20k lines; this
has some smallish effect on performance, but should be safe value
for any modern-ish laptop or desktop.  Could be possible to detect
free memory and work from that, but that is for later; and could
not hold unzipped file in memory, also for later.

- Changed progress messaging to be ProgressBar messaging; don't know
table length so can't do percentage at this point.

- Added in more detail re: bossquery sort, what, and where options to
bin/scripts.rst

dcunning11235 commented:
@dkirkby Have added changes to branch per comments, please review

dkirkby commented:
Btw, you can test updates to the docs using:

```
cd docs
make html
open _build/html/index.html
```

dkirkby commented:
I am testing your branch on my laptop now...   Is `create_meta_full` working on your laptop?

dkirkby commented on bossdata/meta.py:
```diff
@@ -98,36 +100,74 @@ def create_meta_lite(sp_all_path, db_path, verbose=True):
+            # Don't know final count, can't do percentage, bar, etc. Display work done instead
+            widgets = [FormatLabel('Processed: %(value)d lines (in: %(elapsed)s)')]
+            progress_bar = ProgressBar(widgets=widgets, maxval=sys.maxint).start()
+
```
I'm not sure what `maxval=sys.maxint` does here, but I noticed that you also don't call `progress_bar.finish()`.  I guess that's because you don't have any way to measure progress if you don't read the whole table into memory at once?

dcunning11235 replied:
It is just "a big number".  I need to have some value otherwise progress_bar.update throws an error.

I was thinking, however, of a lower-level way to get around this; since we know the input file format is fixed, know the byte length per line, it just simple division to get the row count if the total size of the file is known.

Missing finish() was a mistake, not intentional.

dkirkby replied:
Unfortunately the total number of rows is not known in advance.  Given this constraint, the way you are using progress_bar looks good to me now (except for the missing call to `finish`).

dcunning11235 commented:
I have not, I will run it overnight tonight; have until now not wanted to wait for the download to complete.

dkirkby commented:
The full spAll should "only" take about 20 minutes to download (assuming you have the free space).

dkirkby commented on bossdata/meta.py:
```diff
@@ -98,36 +100,74 @@ def create_meta_lite(sp_all_path, db_path, verbose=True):
     """
+
+    # Hardcode to what I think is (quite) conservative value; smaller == slower.
+    # At 50k see memory peak ~2 GB, incldes holding ~750 MB of source in memory.
```
I am confused by these memory sizes: the input file is only ~115Mb and the final sql file is ~470Mb.  How does this add up to 750Mb in memory at once?

dcunning11235 replied:
This is my (admittedly very rough) accounting, based on looking at memory used by the whole process via top:

I am assuming the gz file is unzipped and in memory; that is roughly 750 Mb.  SQL/DB data shouldn't been in memory, in general, I don't think; or not very much; that is going to disk.  This more or less matches that the memory used by the process when running bottoms out at ~800+ MB.

The remainder (which bounces around a lot but peaks at about 2 GB when I had it doing chunks of 50k records at a time; it does 20k now) must be "the rest of it" including the parsed data/table itself, and whatever magic is doing the parsing.

This points at some additional changes to try out:  writing the data to disk, then reading it back a chunk at a time.  It is hard to understand what all this memory is being used for in parsing, if that is where the rest of it is going; one thing to try here would be to specify the format rather than let it be inferred, and see if this helps

dkirkby commented:
Here are timings running on my laptop:
- original code: 25 mins
- new code: 49 mins

They both give the same answer for the example in `scripts.rst`.

dkirkby commented:
I think the problem is that `astropy.table` does a nice job of automatically inferring the column names and types, but is relatively slow (and memory intensive) at reading the whole file into memory.  On the other hand, `numpy.loadtxt` is relatively fast and memory efficient, but has no automated inference.

However, we can get the best of both worlds.  Can you try this test program:

```
import gzip
import astropy.table
import numpy as np

filename = 'spall.dat.gz'

# Pass-1 uses only the first 100 lines to automatically determine the column names and types.
lines = ''
with gzip.open(filename) as f:
    for i in range(100):
        lines += f.readline()
table = astropy.table.Table.read(lines, format='ascii.fixed_width_two_line', guess=False)

# Pass-2 uses the pass-1 structure to read the whole file.
data = np.loadtxt(filename, skiprows=2, dtype=table.dtype)
print data.nbytes
```

This runs in 72s on my laptop with a memory usage that peaks at 2.5 Gb.  The final in-memory representation of the full db uses only `data.nbytes = 513Mb` (which is smaller than the un-zipped file).

dcunning11235 committed:
Added in missing progress_bar.finish()

dcunning11235 commented:
I replaced most of what I'd done in bossquery with the above and it appears to be working well; peaks at a little over 3 GB, then drops to just over 1350 MB while it does the inserts. (From `top` output)  (This on my VM, which is configured to mimic my laptop right now... actually has slightly less memory.)

Hrmm... grumble :)

On my actual laptop, bossquery immediately dies on attempting to load the HDU 1 from the FITS file, on the line:

```
table = hdulist[1].read()
```

with a MemoryError; it is failing to allocate an array down in fitslib.py.  Maybe I can sort this one out :)

dkirkby commented:
So `create_meta_lite` is working for you now, and now you are tackling `create_meta_full` ?  If so, lets finish this pull request for `create_meta_lite` and start a new issue for `create_meta_full`.

dcunning11235 committed:
Final (hopefully) fixes to meta.py create_meta_lite incorporating dkirkby's changes

dcunning11235 committed:
trivial pre-push cleanup

dcunning11235 commented:
Okay, just the doc changes, the `bossquery --sort` option, and the `create_meta_lite` are now par tof the branch.

dkirkby commented:
Here are the updated timings on my laptop for just creating the sqlite db from the in-memory meta data (so not including any download time):
- original code: 25 mins
- new code: 3 mins 7 secs

Success!

dkirkby commented:
The update docs build fine and look good.
